src/System.Runtime.Extensions/src/System/Environment.Win32.cs

```diff
@@ -2,60 +2,108 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
-using Internal.Runtime.Augments;
-using System.Diagnostics;
 using System.IO;
-using System.Runtime.InteropServices;
 using System.Text;
+using System.Runtime.InteropServices;
 
 namespace System
 {
     public static partial class Environment
     {
-        static partial void GetUserName(ref string username)
+        public static string UserName
         {
-            // Use GetUserNameExW, as GetUserNameW isn't available on all platforms, e.g. Win7
-            var domainName = new StringBuilder(1024);
-            uint domainNameLen = (uint)domainName.Capacity;
-            if (Interop.Secur32.GetUserNameExW(Interop.Secur32.NameSamCompatible, domainName, ref domainNameLen) == 1)
+            get
             {
-                string samName = domainName.ToString();
-                int index = samName.IndexOf('\\');
+                // 40 should be enough as we're asking for the SAM compatible name (DOMAIN\User).
+                // The max length should be 15 (domain) + 1 (separator) + 20 (name) + null. If for
+                // some reason it isn't, we'll grow the buffer.
+
+                // https://support.microsoft.com/en-us/help/909264/naming-conventions-in-active-directory-for-computers-domains-sites-and
+                // https://msdn.microsoft.com/en-us/library/ms679635.aspx
+
+                Span<char> initialBuffer = stackalloc char[40];
+                var builder = new ValueStringBuilder(initialBuffer);
+                GetUserName(ref builder);
+
+                ReadOnlySpan<char> name = builder.AsSpan();
+                int index = name.IndexOf('\\');
                 if (index != -1)
                 {
-                    username = samName.Substring(index + 1);
-                    return;
+                    // In the form of DOMAIN\User, cut off DOMAIN\
+                    name = name.Slice(index + 1);
```

Comment:
If slash is last character somehow this fails gracefully?

Reply:
Good eye, but should never happen- it isn't valid as part of the user name. We didn't check for that case before either.